gflownet/evaluator/base.py

```diff
@@ -0,0 +1,564 @@
+"""
+Base evaluator class for a :class:`~gflownet.gflownet.GFlowNetAgent`.
+
+In charge of evaluating a generic :class:`~gflownet.gflownet.GFlowNetAgent`,
+computing metrics plotting figures and optionally logging results using the
+:class:`~gflownet.gflownet.GFlowNetAgent`'s :class:`~gflownet.utils.logger.Logger`.
+
+Take this :class:`BaseEvaluator` as example to implement your own evaluator class
+for your custom use-case.
+
+.. important::
+
+    Prefer the :meth:`~gflownet.evaluator.abstract.AbstractEvaluator.from_dir`
+    and :meth:`~gflownet.evaluator.abstract.AbstractEvaluator.from_agent`
+    class methods to instantiate an evaluator.
+
+See :ref:`using an evaluator` for more details about how to use an Evaluator.
+"""
+
+import copy
+import pickle
+import time
+from collections import defaultdict
+
+import numpy as np
+import torch
+from scipy.special import logsumexp
+
+from gflownet.evaluator.abstract import ALL_REQS  # noqa
+from gflownet.evaluator.abstract import METRICS  # noqa
+from gflownet.evaluator.abstract import AbstractEvaluator
+from gflownet.utils.batch import Batch
+from gflownet.utils.common import batch_with_rest, tfloat, torch2np
+
+
+class BaseEvaluator(AbstractEvaluator):
```

Comment:
What is the purpose of having a distinct Base and Abstract evaluator? Typically they are one and the same.

Is the idea that the `Base` evaluator is a actually `Default` evaluator, and in future scenarios you imagine someone would need to build some different Evaluator from `Abstract`, but it wouldn't be useful to start by inheriting `Base`?

Reply:
Yes that's it. the nomenclature (`Default` vs `Base`) is just inherited from the rest of the code base and I did contemplate naming it `DefaultEvaluator` instead

Reply:
I imagine most would inherit `Base` not `Abstract` when they do regular stuff but if `Base` does things the wrong way for their use-case (or even too much) they they can just inherit from `Abstract`.

The only requirements are those defined in `Abstract`* because they are exposed to the rest of the code, the methods extending classes (including `Base`) should/can implement are for internal uses.

\* -> `eval_and_log()`, `should_*()`

Reply:
We can totally rename stuff in the rest of the codebase if there would be better choices.